### PR TITLE
poor mans switch panel on tablet

### PR DESCRIPTION
**Pull request by sweiland-openrails**
Basically it's a webpage with 10 times 4 buttons. Buttons for instance for closing doors, direction, switching front light etc. Buttons also contain a label with the status. For doors: closed, opening, open and closing. Blinks red when doors are open and train moves. Hard to implement this with real hardware. I display this webpage on a tablet sitting between the keyboard and monitor.

I'm using the keyboard for brakes, forward, etc. The keys you need every second. Different keys can be shown on the tablet for different types of locomotives: steam, diesel or electric. There are not much buttons for steam engines yet.

Implementing the websocket part is rather clean and not that much code. The same for firing the action in Open Rails when pushing the button on the tablet. However determining the status needs a lot of code, as there is no such thing in Open Rails. But not all buttons would need a status displayed, for instance subwindows like map, track monitor etc. would not need it.

I hope this does not interfere with existing OR web solutions. I saw for instance that César recently created a "Using Arduino devices with Open Rails" change. If it does gives problems undo the PR, no problem.

https://trello.com/c/1EISQgFg/566-poor-mans-switch-panel

### CODE REVIEW

**Comment by cesarBLG** [comment]
That's really nice! It's great that you used WebSockets. I'll study your code to see if we can build an standarized API using them. Now I'm using GET and POST requests which are probably slow for some purposes.

I don't think my changes will conflict with yours, but I'd suggest you take advantage from them. I created the ExternalDeviceState and ExternalDeviceButton classes, which already manage IsPressed, IsUp, IsDown, etc. for you. You could inherit your SwitchPanel from them, which would unify all devices. Then, instead of adding SwitchPanelModule.IsPressed() to MSTSLocomotiveViewer, you would add it to the UserInput.IsPressed function. You can follow the example of the RailDriver class, which I refactored so that RailDriver and WebDeviceState share common code.

**Comment by sweiland-openrails** [comment]
Thanks for looking into my changes. 

Yes, I think websockets are faster and more reliable. But one would need a websocket implementation for the Arduino. Looks like that's available. Also for the ESP32, my preferred MCU as it has WIFI onboard.

I'm not so good with GIT, so I might wait with your suggestions until after my enhancement is merged, but I'll give it a try. Should I pull master into my local branch? So that I see your changes you made in PR 715 locally and can react on that?

**Comment by cesarBLG** [comment]
> Yes, I think websockets are faster and more reliable. But one would need a websocket implementation for the Arduino. Looks like that's available. Also for the ESP32, my preferred MCU as it has WIFI onboard.

I'm using a bridge program which converts information from the OR interface to serial port data, so this is not a problem. Websockets would probably add too much overhead for the MCU, and they cannot be used with non-Wifi boards. Using an Arduino Mega + Serial Port bridge is great because it's cheap but you get lots of GPIO, which is what you need for a cab desk. 

> I'm not so good with GIT, so I might wait with your suggestions until after my enhancement is merged, but I'll give it a try. Should I pull master into my local branch? So that I see your changes you made in PR 715 locally and can react on that?

You should update your master branch using "git pull" (you have to ensure that you pull from openrails/openrails and not from your fork sweiland-openrails/openrails, otherwise you won't update anything). Then you switch to the SwitchPanel branch and perform a "git merge master"

It is usually a good idea that you update your master branch before creating a new branch from it. This allows you to start working from the most recent codebase, and will reduce the possiblity of conflicting with other changes.

In this case my suggestion of merging my changes is not because of conflicts (I think there aren't any, since the Unstable version contains your panel), but because I think it would simplify the code a bit.

I'm surprised that you had to create a function for every control you're inserting to the panel. This means that whenever we add a new control, the interface would need to be updated. Do you think there is a solution that gets all data automatically?

**Comment by sweiland-openrails** [comment]
> I'm using a bridge program which converts information from the OR interface to serial port data, so this is not a problem. Websockets would probably add too much overhead for the MCU, and they cannot be used with non-Wifi boards. Using an Arduino Mega + Serial Port bridge is great because it's cheap but you get lots of GPIO, which is what you need for a cab desk.

Yes, I know. For my modeltrains I use two atmega32's for controlling the 2*28 leds on my layout. Leds in houses switched on/off randomly etc. Connected to a rpi zero for programming and controlling.
> You should update your master branch using "git pull" (you have to ensure that you pull from openrails/openrails and not from your fork sweiland-openrails/openrails, otherwise you won't update anything). Then you switch to the SwitchPanel branch and perform a "git merge master"

succeeded, thanks.

> I'm surprised that you had to create a function for every control you're inserting to the panel. This means that whenever we add a new control, the interface would need to be updated. Do you think there is a solution that gets all data automatically?

The controls on this panel are only a subset. I would not expect every new control has to be added. And needing custom code because first of all you need to place the button in the 4 times 10 matrix. And needing a description, which cannot be found in Open Rails. And for some controls I have two buttons instead of one. For instance gear up and down. However, I have enhanced this code a bit. I removed the extra function in between. And created a default function for most buttons, grabbing the description  from the UserCommand enum. I hope this works ok for translations, have not yet seen a translated Open Rails, should gather some knowledge about this. For retrieving the status  there is also much code.

> I don't think my changes will conflict with yours, but I'd suggest you take advantage from them. I created the ExternalDeviceState and ExternalDeviceButton classes, which already manage IsPressed, IsUp, IsDown, etc. for you. You could inherit your SwitchPanel from them, which would unify all devices.

Code does not conflict. In my last commit I removed the IsPressed, only the IsUp and IsDown are left. I did not succeed in enhancing my code with your suggestion. That's yet too complicated for me. I might learn enough in the future to give it another try. Back in 1981 when I started programming as a career there was no object oriented programming... just Cobol.

**Comment by cesarBLG** [comment]
> The controls on this panel are only a subset. I would not expect every new control has to be added. And needing custom code because first of all you need to place the button in the 4 times 10 matrix. And needing a description, which cannot be found in Open Rails. And for some controls I have two buttons instead of one. For instance gear up and down. However, I have enhanced this code a bit. I removed the extra function in between. And created a default function for most buttons, grabbing the description from the UserCommand enum. I hope this works ok for translations, have not yet seen a translated Open Rails, should gather some knowledge about this. For retrieving the status there is also much code.

I see, thanks. It's great that you managed to add a default function for most buttons.

> Code does not conflict. In my last commit I removed the IsPressed, only the IsUp and IsDown are left. I did not succeed in enhancing my code with your suggestion. That's yet too complicated for me. I might learn enough in the future to give it another try. Back in 1981 when I started programming as a career there was no object oriented programming... just Cobol.

Don't worry, I'll take a look at it later. Why are you using IsDown/IsUp here instead of IsPressed/IsReleased? The code expects the command to be triggered only once, but IsDown/IsUp return true as long as the relevant button is kept pressed/released
```csharp
UserInput.IsPressed(command) || SwitchPanelModule.IsDown(command)
```

**Comment by sweiland-openrails** [comment]
> Why are you using IsDown/IsUp here instead of IsPressed/IsReleased?

At first for every key I decided whether it needed a IsPressed (for instance Control Door) or Isdown/IsUp. The last one was to be used for instance for the Sander button, which should be on as long as the button is pressed. Then I decided to test the IsDown/IsUp for both situations, reasoning that the IsDown is more or less the same as the IsPressed. And it worked, at least with the way I coded the javascript for the tablet. So less code leaving out the IsPressed. Might be I use the wrong names. In javascript the events are touchstart and touchend.

**Comment by cesarBLG** [comment]
You should always use IsPressed/IsReleased for any type of control, regardless of whether it has to be continuously pressed (e.g. Sander) or not (e.g. pantograph). That is handled internally in the Input code. Your SwitchPanel should implement the following methods:

- IsPressed: returns true only once, just after the button is pressed
- IsReleased: returns true only once, just after the button is released
- IsDown: returns true as long as the button is kept pressed

Also, you don't need to add SwitchPanelModule.IsDown(command) here
```csharp
if (UserInput.IsPressed(command) || SwitchPanelModule.IsDown(command))
```
because UserInput.IsPressed() already contains a call for SwitchPanelModule.IsDown(), so this is redundant.

**Review by cjakeman — COMMENTED** [review]
Hi Siebren,

It's great to see this new feature taking shape.

Before PRs can be approved, they need either a Trello card or a Launchpad Blueprint to be created and approved. If you can create one and add a link to it in this PR, then I can approve it.

Would you also add to this PR a link to the forum discussion for this enhancement, please.

Thanks,

Chris

**Comment by sweiland-openrails** [comment]
hi Chris,

correct me if I'm wrong, but I was under the impression that a pull request made it into the Unstable version. So that people could take a look and comment before he final merge. Maybe that was not the case because it was still a draft pull request and removed the draft from it. But on the latest unstable "Open Rails U2023.06.11-0752 2023-06-11 08-58-14" I do not see my "switchpanel".

Added Trello card. Elvastower discussion in http://www.elvastower.com/forums/index.php?/topic/37038-poor-mans-switch-panel/page__p__296016#entry296016.

regards, Siebren.

**Comment by cjakeman** [comment]
Hi Siebren, 

> correct me if I'm wrong, but I was under the impression that a pull request made it into the Unstable version. So that people could take a look and comment before he final merge. 

Spot on, but the output from the process that merges PRs into the Unstable version can be found at https://james-ross.co.uk/projects/or
You can see from this that your PR 831 conflicts with changes from another PR and so your PR was excluded from the build of Unstable.

If you need help in fixing this, then I'll see what I can do.

> Added Trello card. Elvastower discussion in http://www.elvastower.com/forums/index.php?/topic/37038-poor-mans-switch-panel/page__p__296016#entry296016.

Thanks for adding the card. I have approved it by moving it from the Unsorted list to the 1.6 list.

Best wishes,

Chris

**Comment by sweiland-openrails** [comment]
hi Chris,

thanks for pointing me to the right direction. The problem was in file Source\ORTS.Common\Input\UserCommand.cs, where in PR 757 "Unify RailDriver code implementations" a line was added. In my PR all lines are changed, I added:

`[GetParticularString("SwitchPanel", "Autopilot")] `

in every enum value (empty string if not used in SwitchPanel). Git merge could not handle both changes, tried some changes in my version, but no. I had to remove my changes until after the added line by PR 751. Don't know if it is wise to pollute the file UserCommand.cs with these extra SwitchPanel strings. Seemed a logical way to code, but can be done easily elsewhere in one of the SwitchPanel files. Please let me know what you think.

The unstable version now contains the SwitchPanel. I'll add some more buttons for the steam locomotives.

regards, Siebren.

**Comment by sweiland-openrails** [comment]
Moved the descriptions for the Switchpanel buttons from UserCommnd.cs to one of the SwitchPanel sources. I thinks it's better this way, leaving UserCommand.cs untouched. Also seen the merge problems in UserCommand.cs.

**Comment by sweiland-openrails** [comment]
dear César,

Would you find some time to review this enhancement? As you have already looked into into the code. I've asked Mauricio also. As he was mentioning this PR in a welcome Elvas Tower message. Still trying to find my way into the code and the review process. Don't hesitate to tell if I'm doing something wrong.

I did not yet have any comment from users testing the unstable version. Do not know why.

regards, Siebren.

**Comment by cesarBLG** [comment]
Sure, I'll review it as soon as possible. I have no idea about JavaScript, but I'll try my best.

File: Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchOnPanelDefinition.cs
Comment (cesarBLG) [inline Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchOnPanelDefinition.cs:329]: I'm curious, how do you set those numbers?

File: Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchesOnPanel.cs
Comment (cesarBLG) [inline Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchesOnPanel.cs:53]: Same comment about casing

**Review by cesarBLG — CHANGES_REQUESTED** [review]
I have a couple of suggestions, but nothing really serious.

Trello card isn't sorted, so I cannot approve yet.

File: Source/RunActivity/Viewer3D/RollingStock/MSTSLocomotiveViewer.cs
Comment (sweiland-openrails) [reply Source/RunActivity/Viewer3D/RollingStock/MSTSLocomotiveViewer.cs]: good point, done

File: Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchPanelModule.cs
Comment (sweiland-openrails) [reply Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchPanelModule.cs]: Changed as suggested. Think I followed the naming of the javascript events: mousedown/mouseup. But this is better.

File: Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchOnPanelDefinition.cs
Comment (sweiland-openrails) [reply Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchOnPanelDefinition.cs]: Hope I've changed all the lower case function names

File: Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchOnPanelDefinition.cs
Comment (sweiland-openrails) [reply Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchOnPanelDefinition.cs:166]: Yes, it might be part of the enum UserCommand, however then it ends up into another assembly. On Elvis Tower I tried to start a discussion for this problem: http://www.elvastower.com/forums/index.php?/topic/37156-translations-gnugettext/page__p__297487__hl__translations__fromsearch__1#entry297487. But got no answers. But I think the current implementation is ok.

File: Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchOnPanelDefinition.cs
Comment (sweiland-openrails) [reply Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchOnPanelDefinition.cs:329]: Without this function I get the VS Compiler Warning (level 3) CS0659:
'class' overrides Object.Equals(object o) but does not override Object.GetHashCode()
To get rid of this warning you can have VS generate this function. If doing so, it generates another number every time. 
This function ObjectGetHashCode is not that important. But for new sources I try to suppress every warning.

File: Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchesOnPanel.cs
Comment (sweiland-openrails) [reply Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchesOnPanel.cs]: done

**Comment by sweiland-openrails** [comment]
hi César,

thanks for the review. See my comments for the details above.

What do I have to do with the Trello card? And do these Trello cards get replaced with something from Github in the future?
And do I have to push the "Resolve conversation" button or is that something for the reviewer?
And as a developer do you look at the "Codacy Static Code Analysis" and "CodeFactor" issues? I do not find them that relevant...

regards, Siebren.

File: Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchOnPanelDefinition.cs
Comment (cesarBLG) [reply Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchOnPanelDefinition.cs:329]: Thanks for the explanation!

File: Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchOnPanelDefinition.cs
Comment (cesarBLG) [reply Source/RunActivity/Viewer3D/WebServices/SwitchPanel/SwitchOnPanelDefinition.cs:166]: Okay, I'm no expert either, so let's keep it as is

**Comment by cesarBLG** [comment]
> What do I have to do with the Trello card? And do these Trello cards get replaced with something from Github in the future?

You don't have to do any additional step. ORMT usually approves new features after forum discussion with positive feedback, but maybe they forgot about this feature (it happens from time to time). You can contact them directly to get it approved (e.g. in the "ORMT-Sessions" thread.

> And do I have to push the "Resolve conversation" button or is that something for the reviewer?

You can resolve them if you don't expect new messages (e.g. because you have already tackled the issues mentioned by the reviewer). But it doesn't matter, it can be easily undone if necessary.

> And as a developer do you look at the "Codacy Static Code Analysis" and "CodeFactor" issues? I do not find them that relevant...

I only look at them just in case they detect silly bugs like doing a = 5 instead of a == 5 (it happened to me and I noticed thanks to that), but otherwise I don't look at them. In fact I'm against some of its suggestions, which in my opinion make the code harder to read.

**Comment by sweiland-openrails** [comment]
hi Chris,

can you sort the Trello card for this enhancement? As César has reviewed this but is not able to approve: "Trello card isn't sorted, so I cannot approve yet."

thanks, Siebren.

**Comment by cjakeman** [comment]
Hi Siebren,

 

Thanks for getting in touch. I’m sorry about this PR - I thought I had already sorted the card. It’s done now.

 

 

Best wishes,

 

Chris

**Review by cesarBLG — APPROVED** [review]
Thanks for addressing my comments, it looks good to me now

**Pull request merged**
